src/query_plan/operation.rs

```diff
@@ -1247,6 +1261,42 @@ impl NormalizedSelectionSet {
     ) -> Result<NormalizedSelectionSet, FederationError> {
         todo!()
     }
+
+    pub(crate) fn add_at_path(
+        &mut self,
+        path: &OpPath,
+        selection_set: Option<&Arc<NormalizedSelectionSet>>,
+    ) {
+        Arc::make_mut(&mut self.selections).add_at_path(path, selection_set)
+    }
+
+    pub(crate) fn add(&mut self, _selections: &NormalizedSelectionSet) {
+        todo!()
+    }
+}
+
+impl NormalizedSelectionMap {
+    /// Adds a path, and optional some selections following that path, to those updates.
+    ///
+    /// The final selections are optional (for instance, if `path` ends on a leaf field,
+    /// then no followup selections would make sense),
+    /// but when some are provided, uncesssary fragments will be automaticaly removed
+    /// at the junction between the path and those final selections.
+    /// For instance, suppose that we have:
+    ///  - a `path` argument that is `a::b::c`,
+    ///    where the type of the last field `c` is some object type `C`.
+    ///  - a `selections` argument that is `{ ... on C { d } }`.
+    /// Then the resulting built selection set will be: `{ a { b { c { d } } }`,
+    /// and in particular the `... on C` fragment will be eliminated since it is unecesasry
+    /// (since again, `c` is of type `C`).
+    pub(crate) fn add_at_path(
```

Comment:
As I understand it the point of that module is to keep `IndexMap<NormalizedSelectionKey, NormalizedSelection>` private and expose and API like a set of `NormalizedSelection`s. Code inside the module is responsible for keeping `NormalizedSelectionKey`s consistent with their values. Code outside of it doesn’t see keys at all so we don’t need to worry about getting them wrong. It can rely on lower-level methods like `insert`, which are inside of the module.